File: src/python/virtual_fs.py
"""In-memory virtual filesystem with lazy file support."""

from __future__ import annotations

import copy
import fnmatch
import os
from typing import Any, Callable


class VirtualFS:
    """Simple in-memory filesystem. Paths are always absolute and normalized."""

    def __init__(self, files: dict[str, str | bytes] | None = None):
        self._files: dict[str, str | bytes] = {}
        # Lazy file providers — called on first read, result cached
        self._lazy: dict[str, Callable[[], str | bytes]] = {}
        if files:
            for path, content in files.items():
                self.write(self._norm(path), content)

    @staticmethod
    def _norm(path: str) -> str:
        result = os.path.normpath("/" + path)
        # os.path.normpath preserves leading // on POSIX — collapse it
        if result.startswith("//"):
            result = result[1:]
        return result

    def write(self, path: str, content: str | bytes) -> None:
        path = self._norm(path)
        self._files[path] = content

    def write_lazy(self, path: str, provider: Callable[[], str | bytes]) -> None:
        """Register a lazy file — provider called on first read, then cached."""
        path = self._norm(path)
        self._lazy[path] = provider

    def read(self, path: str) -> str | bytes:
        path = self._norm(path)
        # Resolve lazy providers
        if path in self._lazy:
            self._files[path] = self._lazy.pop(path)()
        if path not in self._files:
            raise FileNotFoundError(f"{path}: No such file")
        return self._files[path]

    def read_text(self, path: str) -> str:
        content = self.read(path)
        return content if isinstance(content, str) else content.decode("utf-8", errors="replace")

    def exists(self, path: str) -> bool:
        path = self._norm(path)
        return path in self._files or path in self._lazy or self._is_dir(path)

    def remove(self, path: str) -> None:
        path = self._norm(path)
        if path in self._files:
            del self._files[path]
        elif path in self._lazy:
            del self._lazy[path]
        else:
            raise FileNotFoundError(f"{path}: No such file")

    def _all_paths(self) -> set[str]:
        return set(self._files.keys()) | set(self._lazy.keys())

    def _is_dir(self, path: str) -> bool:
        path = self._norm(path)
        prefix = path.rstrip("/") + "/"
        return any(p.startswith(prefix) for p in self._all_paths())

    def listdir(self, path: str = "/") -> list[str]:
        path = self._norm(path).rstrip("/") + "/"
        if path == "//":
            path = "/"
        entries: set[str] = set()
        for p in self._all_paths():
            if p.startswith(path) and p != path:
                rest = p[len(path):]
                entry = rest.split("/")[0]
                entries.add(entry)
        return sorted(entries)

    def find(self, root: str = "/", pattern: str = "*") -> list[str]:
        root = self._norm(root).rstrip("/")
        results = []
        for p in sorted(self._all_paths()):
            if not p.startswith(root):
                continue
            basename = p.rsplit("/", 1)[-1]
            if fnmatch.fnmatch(basename, pattern):
                results.append(p)
        return results

    def stat(self, path: str) -> dict[str, Any]:
        path = self._norm(path)
        if self._is_dir(path):
            return {"path": path, "type": "directory"}
        content = self.read(path)
        size = len(content.encode("utf-8") if isinstance(content, str) else content)
        return {"path": path, "type": "file", "size": size}

    def clone(self) -> VirtualFS:
        """Create an independent copy of this filesystem."""
        new = VirtualFS()
        new._files = copy.deepcopy(self._files)
        new._lazy = dict(self._lazy)
        return new

File: src/typescript/virtual-fs.ts
/**
 * In-memory virtual filesystem with lazy file support.
 * String-only content (no bytes).
 */

export class VirtualFS {
  private _files: Map<string, string> = new Map();
  private _lazy: Map<string, () => string> = new Map();
  private _mtimes: Map<string, number> = new Map();
  private _mtimeCounter: number = 0;

  constructor(files?: Record<string, string>) {
    if (files) {
      for (const [path, content] of Object.entries(files)) {
        this.write(path, content);
      }
    }
  }

  private _stamp(path: string): void {
    this._mtimeCounter += 1;
    this._mtimes.set(path, this._mtimeCounter);
  }

  static _norm(path: string): string {
    if (!path.startsWith("/")) {
      path = "/" + path;
    }
    const parts = path.split("/");
    const resolved: string[] = [];
    for (const part of parts) {
      if (part === "" || part === ".") continue;
      if (part === "..") {
        resolved.pop();
      } else {
        resolved.push(part);
      }
    }
    return "/" + resolved.join("/");
  }

  write(path: string, content: string): void {
    path = VirtualFS._norm(path);
    this._files.set(path, content);
    this._lazy.delete(path);
    this._stamp(path);
  }

  writeLazy(path: string, provider: () => string): void {
    path = VirtualFS._norm(path);
    this._lazy.set(path, provider);
    this._stamp(path);
  }

  read(path: string): string {
    path = VirtualFS._norm(path);
    if (this._lazy.has(path)) {
      const provider = this._lazy.get(path)!;
      this._lazy.delete(path);
      const content = provider();
      this._files.set(path, content);
    }
    const content = this._files.get(path);
    if (content === undefined) {
      throw new Error(`${path}: No such file`);
    }
    return content;
  }

  exists(path: string): boolean {
    path = VirtualFS._norm(path);
    return this._files.has(path) || this._lazy.has(path) || this._isDir(path);
  }

  remove(path: string): void {
    path = VirtualFS._norm(path);
    if (this._files.has(path)) {
      this._files.delete(path);
    } else if (this._lazy.has(path)) {
      this._lazy.delete(path);
    } else {
      throw new Error(`${path}: No such file`);
    }
    this._mtimes.delete(path);
  }

  private _allPaths(): Set<string> {
    const paths = new Set<string>(this._files.keys());
    for (const k of this._lazy.keys()) {
      paths.add(k);
    }
    return paths;
  }

  _isDir(path: string): boolean {
    path = VirtualFS._norm(path);
    const prefix = path.replace(/\/+$/, "") + "/";
    for (const p of this._allPaths()) {
      if (p.startsWith(prefix)) return true;
    }
    return false;
  }

  listdir(path: string = "/"): string[] {
    path = VirtualFS._norm(path).replace(/\/+$/, "") + "/";
    if (path === "//") path = "/";
    const entries = new Set<string>();
    for (const p of this._allPaths()) {
      if (p.startsWith(path) && p !== path) {
        const rest = p.slice(path.length);
        const entry = rest.split("/")[0];
        entries.add(entry);
      }
    }
    return [...entries].sort();
  }

  find(root: string = "/", pattern: string = "*"): string[] {
    root = VirtualFS._norm(root).replace(/\/+$/, "");
    const regex = VirtualFS._globToRegex(pattern);
    const results: string[] = [];
    const allPaths = [...this._allPaths()].sort();
    for (const p of allPaths) {
      if (!p.startsWith(root)) continue;
      const basename = p.split("/").pop() || "";
      if (regex.test(basename)) {
        results.push(p);
      }
    }
    return results;
  }

  private static _globToRegex(pattern: string): RegExp {
    let regexStr = "^";
    for (const ch of pattern) {
      if (ch === "*") regexStr += ".*";
      else if (ch === "?") regexStr += ".";
      else if (".+^${}()|[]\\".includes(ch)) regexStr += "\\" + ch;
      else regexStr += ch;
    }
    regexStr += "$";
    return new RegExp(regexStr);
  }

  stat(path: string): { path: string; type: string; size?: number; mtime?: number } {
    path = VirtualFS._norm(path);
    if (this._isDir(path)) {
      return { path, type: "directory" };
    }
    const content = this.read(path);
    const size = new TextEncoder().encode(content).length;
    return { path, type: "file", size, mtime: this._mtimes.get(path) ?? 0 };
  }

  clone(): VirtualFS {
    const newFs = new VirtualFS();
    for (const [k, v] of this._files) {
      newFs._files.set(k, v);
    }
    for (const [k, v] of this._lazy) {
      newFs._lazy.set(k, v);
    }
    for (const [k, v] of this._mtimes) {
      newFs._mtimes.set(k, v);
    }
    newFs._mtimeCounter = this._mtimeCounter;
    return newFs;
  }
}
